### PR TITLE
Make runners configurable from the command-line using `hydra`

### PR DESCRIPTION
**Pull request by charlesincharge**
https://github.com/agencyenterprise/neural-data-simulator/issues/44

#### Introduction

This PR adds command-line configuration to the `run_*.py` scripts. For example you can now change the logger level with:
`streamer log_level=DEBUG`.

#### Changes

* Switch from [`pydantic-yaml`](https://github.com/NowanIlfideme/pydantic-yaml) library to [`hydra`](https://hydra.cc/docs/intro/). `hydra` both reads in YAML config files and parses any values from the command-line interface
    * The config files (mostly) did not need to change for this update.
* Replaces deprecated `YamlStrEnum` by `str, Enum` mix-inheritance

This hydra/PyDantic integration is based on this blog post: https://suneeta-mall.github.io/2022/03/15/hydra-pydantic-config-management-for-training-application.html#can-pydantic-bridge-the-gaps-in-hydra

#### Behavior

* `run_closed_loop` works the same: 
Screen Recording 2023-10-11 at 4 23 19 PM 2
* The `run_*.py` scripts can now be configured on the command-line, e.g. `streamer log_level=DEBUG streamer.stream_indefinitely=false`
* `run_*.py --help` shows the default config, e.g.:
```❯ streamer --help
run_streamer is powered by Hydra.

== Configuration groups ==
Compose your configuration from those groups (group=option)



== Config ==
Override anything in the config (foo.bar=value)

version: 1.0.0
log_level: INFO
streamer:
  lsl_chunk_frequency: 100
  input_type: npz
  stream_indefinitely: true
  blackrock:
    input:
      file: sample_data/blackrock-file.ns6
    output:
      lsl:
        channel_format: int16
        instrument:
          manufacturer: Blackrock Neurotech
          model: Playback
          id: 1
  npz:
    input:
      file: sample_data/session_4_behavior_standardized.npz
      timestamps_array_name: timestamps_train
      data_array_name: vel_train
    output:
      sampling_rate: 50
      n_channels: 2
      lsl:
        channel_format: float32
        stream_name: NDS-Behavior
        stream_type: Behavior
        source_id: SimulatedBehavior
        instrument:
          manufacturer: Blackrock Neurotech
          model: Simulated
          id: 0
```
* The scripts do also generate an `outputs` folder, which is mostly unused right now.

### CODE REVIEW

**Comment by charlesincharge** [comment]
@NewtonSander have you seen this error before? https://github.com/agencyenterprise/neural-data-simulator/actions/runs/6583346051/job/17886339278?pr=49

On `(test) windows-latest`:
```
Run poetry install --no-interaction --no-root

Command ['D:\\a\\neural-data-simulator\\neural-data-simulator\\.venv\\Scripts\\python.exe', '-I', '-W', 'ignore', '-'] errored with the following return code 103

Error output:
No Python at '"C:\hostedtoolcache\windows\Python\3.[11](https://github.com/agencyenterprise/neural-data-simulator/actions/runs/6583346051/job/17886339278?pr=49#step:8:12).5\x64\python.exe'
```

My PR doesn't touch any actions, so I'm a little confused where this is coming from, unless `windows-latest` was updated and it broke our tests somehow?

**Comment by NewtonSander** [comment]
> @NewtonSander have you seen this error before? https://github.com/agencyenterprise/neural-data-simulator/actions/runs/6583346051/job/17886339278?pr=49
> 
> On `(test) windows-latest`:
> 
> ```
> Run poetry install --no-interaction --no-root
> 
> Command ['D:\\a\\neural-data-simulator\\neural-data-simulator\\.venv\\Scripts\\python.exe', '-I', '-W', 'ignore', '-'] errored with the following return code 103
> 
> Error output:
> No Python at '"C:\hostedtoolcache\windows\Python\3.[11](https://github.com/agencyenterprise/neural-data-simulator/actions/runs/6583346051/job/17886339278?pr=49#step:8:12).5\x64\python.exe'
> ```
> 
> My PR doesn't touch any actions, so I'm a little confused where this is coming from, unless `windows-latest` was updated and it broke our tests somehow?

@charlesincharge I wasn't aware of this error, I re-triggered a successful job on `main` and it also failed.
I've tried to fix the bug in a fork of main and didn't manage to find the root cause, so the workaround was to disable the cache for Windows:
https://github.com/agencyenterprise/neural-data-simulator/pull/53

File: docs/source/decoders.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ The {class}`decoder.decoders.Decoder` class can also be used with a custom decod
 
 ## Configuring and running the included decoder
 
-To configure the decoder, change the file `settings_decoder.yaml`, which is located by default in the `$HOME/.nds/` folder. You can point the script to use a different configuration file by passing the `--settings-path` argument.
+To configure the decoder, change the file `settings_decoder.yaml`, which is located by default in the `$HOME/.nds/` folder. You can point the script to use a different configuration file by passing the `--config` argument.
```
Comment (NewtonSander) [inline docs/source/decoders.md:15]: is this parameter `--config` correct? Shouldn't it be `--config-path`?

File: docs/source/ephys_generator.md [hunk 2]
```diff
@@ -55,7 +55,11 @@ To start the `ephys generator`, complete the [installation](installation.md), ad
 ephys_generator
 ```
 
-The script can use a different config file specified via the `--settings-path` argument.
+To use a specific config file, specify the config directory (`--config-dir`) and config file-name (`--config-name`) flags. For example:
+
+```
+ephys_generator --config-dir $HOME/.nds/ --config-name settings
```
Comment (NewtonSander) [inline docs/source/ephys_generator.md:61]: have you tried using a --config-dir different from the default NDS_HOME path (os.path.join(os.path.expanduser("~"), ".nds"))? 
I think it won't work as the runtime.py has the `NDS_HOME` variable that is used in the hydra decorators but it's not set by the user input

**Review by NewtonSander — COMMENTED** [review]
Great job Charles!
The only thing I am missing is the output we had for each command line script when we were passing the `--help` flag.

File: docs/source/decoders.md (hunk 1 above)
Comment (charlesincharge) [reply docs/source/decoders.md]: Thanks for the catch! Updated to `--config-dir` and `--config-name`. Strangely `--config-path` is just an alias for `--config-dir`.

File: docs/source/ephys_generator.md (hunk 2 above)
Comment (charlesincharge) [reply docs/source/ephys_generator.md:61]: Thanks for catching that.

Looks like `--config-path` is the correct way to overwrite the config, whereas `--config-dir` is only appends to the config-search-path. So something like:
```
streamer --config-path=/Users/charles/Development/nds/src/streamer/config/
```
still works after removing my `~/.nds/*.yaml` configs.

Unfortunately, it requires an absolute path (or a path-relative to `streamer`, which is probably a bit confusing for most users).

Let me see if there is a way to also pass in a config path relative to the current working directory.

**Comment by charlesincharge** [comment]
Thanks for helping me workaround that test error Newton!

**Comment by NewtonSander** [comment]
@charlesincharge it seems like I can't run `decoder --help` if the config files do not exist in the default location:
```
decoder --help                    
Traceback (most recent call last):
  File "/Users/nsander/workspace/neural-data-simulator/src/decoder/run_decoder.py", line 110, in run
    run_with_config()
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/main.py", line 94, in decorated_main
    _run_hydra(
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/_internal/utils.py", line 363, in _run_hydra
    hydra.app_help(config_name=config_name, args_parser=args_parser, args=args)
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/_internal/hydra.py", line 356, in app_help
    cfg = self.compose_config(
          ^^^^^^^^^^^^^^^^^^^^
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/_internal/hydra.py", line 594, in compose_config
    cfg = self.config_loader.load_configuration(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/_internal/config_loader_impl.py", line 142, in load_configuration
    return self._load_configuration_impl(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/_internal/config_loader_impl.py", line 243, in _load_configuration_impl
    self.ensure_main_config_source_available()
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/_internal/config_loader_impl.py", line 129, in ensure_main_config_source_available
    self._missing_config_error(
  File "/Users/nsander/.pyenv/versions/3.11.6/lib/python3.11/site-packages/hydra/_internal/config_loader_impl.py", line 102, in _missing_config_error
    raise MissingConfigException(
hydra.errors.MissingConfigException: Primary config directory not found.
Check that the config directory '/Users/nsander/.nds' exists and readable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/nsander/.pyenv/versions/3.11.6/bin/decoder", line 6, in <module>
    sys.exit(run())
             ^^^^^
  File "/Users/nsander/workspace/neural-data-simulator/src/decoder/run_decoder.py", line 112, in run
    raise FileNotFoundError(
FileNotFoundError: Run 'nds_post_install_config' to copy the default settings files
```
after I execute `nds_post_install_config` I get the yaml config and at the end this message:
```
Powered by Hydra (https://hydra.cc)
Use --hydra-help to view Hydra specific help
```

so i have to execute --hydra-help to get the help docs for `--config-path` and `--config-name`, but it's mixed with all the other options provided by hydra:
```
decoder --hydra-help
Hydra (1.3.2)
See https://hydra.cc for more info.

== Flags ==
--help,-h : Application's help
--hydra-help : Hydra's help
--version : Show Hydra's version and exit
--cfg,-c : Show config instead of running [job|hydra|all]
--resolve : Used in conjunction with --cfg, resolve config interpolations before printing.
--package,-p : Config package to show
--run,-r : Run a job
--multirun,-m : Run multiple jobs with the configured launcher and sweeper
--shell-completion,-sc : Install or Uninstall shell completion:
    Bash - Install:
    eval "$(decoder -sc install=bash)"
    Bash - Uninstall:
    eval "$(decoder -sc uninstall=bash)"

    Fish - Install:
    decoder -sc install=fish | source
    Fish - Uninstall:
    decoder -sc uninstall=fish | source

    Zsh - Install:
    Zsh is compatible with the Bash shell completion, see the [documentation](https://hydra.cc/docs/1.2/tutorials/basic/running_your_app/tab_completion#zsh-instructions) for details.
    eval "$(decoder -sc install=bash)"
    Zsh - Uninstall:
    eval "$(decoder -sc uninstall=bash)"

--config-path,-cp : Overrides the config_path specified in hydra.main().
                    The config_path is absolute or relative to the Python file declaring @hydra.main()
--config-name,-cn : Overrides the config_name specified in hydra.main()
--config-dir,-cd : Adds an additional config dir to the config search path
--experimental-rerun : Rerun a job from a previous config pickle
--info,-i : Print Hydra information [all|config|defaults|defaults-tree|plugins|searchpath]
Overrides : Any key=value arguments to override config values (use dots for.nested=overrides)

== Configuration groups ==
Compose your configuration from those groups (For example, append hydra/job_logging=disabled to command line)

hydra: config
hydra/env: default
hydra/help: default
hydra/hydra_help: default
hydra/hydra_logging: colorlog, default, disabled, hydra_debug, none
hydra/job_logging: colorlog, default, disabled, none, stdout
hydra/launcher: basic
hydra/output: default
hydra/sweeper: basic


Use '--cfg hydra' to Show the Hydra config
```

**Comment by charlesincharge** [comment]
After some discussion with Newton and trying to brainstorm workarounds, I have decided that `hydra` is probably over-featured and a bit too opinionated for our use-case. For example, `hydra` is not designed for the situation where the default config file doesn't exist (as it does when the user hasn't run `nds_post_install_config`)

We do want `hydra`'s specific YAML/command-line-merging logic though, and this is implemented by `OmegaConf`. 
I'll change this PR to use https://omegaconf.readthedocs.io/en/2.3_branch/usage.html#from-command-line-arguments

**Comment by charlesincharge** [comment]
@NewtonSander thanks for the feedback here. I made some substantial changes, so I moved it to a different branch and opened a separate PR: https://github.com/agencyenterprise/neural-data-simulator/pull/57

I will close this PR.

**Pull request closed without merging**